File: baekjoon/dp/11051/coefficient.py
import sys

sys.setrecursionlimit(1001)


def get_comb(N, R):
    global memory

    if R == 1:
        return N

    if R == 0 or N == R:
        return 1

    key = (N, R)

    if key in memory:
        return memory[key]

    memory[key] = (((get_comb(N - 1, R - 1) % DIVISOR)
                    + (get_comb(N - 1, R) % DIVISOR)) % DIVISOR)
    return memory[key]


if __name__ == '__main__':
    DIVISOR = 10007
    N, R = map(int, sys.stdin.readline().split())
    memory = dict()

    print(get_comb(N, R))

File: baekjoon/dp/11051/coefficient.ts
const DIVISOR = 10007;
const [N, R] = require('fs').readFileSync(0).toString().split(' ').map(Number);
var dp: Array<Array<number>> = new Array(N + 1);

dp[0] = new Array<number>(R + 1).fill(0);
for (let n = 1; n <= N; ++n)
{
    dp[n] = new Array<number>(R + 1).fill(1);
    for (let r = 1; r <= R; ++r)
    {
        if (n == r)
            continue ;

        dp[n][r] = ((dp[n - 1][r - 1] % DIVISOR)
                    + (dp[n - 1][r] % DIVISOR)) % DIVISOR;
    }
}

console.log(dp[N][R]);
